### PR TITLE
[new-lair-7] lair server abstraction and implementation

### PR DESCRIPTION
**Pull request by neonphog**
DEPENDS ON https://github.com/holochain/lair/pull/65

[Video Walkthrough (2:47)](https://www.youtube.com/watch?v=U5C3AGlZ6QI)

Defines an abstraction for implementing a server, as well as a concrete server implementation that is agnostic to the storage backend and how the server actually manages to communicate with the client (so long as it uses an `AsyncRead / AsyncWrite` channel).

### CODE REVIEW

File: crates/lair_keystore_api/src/lair_server/priv_srv.rs [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+
+        // spawn a task for reading incoming messages from the client
+        tokio::task::spawn(async move {
+            let inner = &inner;
+            let send = &send;
+            let enc_ctx_key = &enc_ctx_key;
+            let dec_ctx_key = &dec_ctx_key;
+            let unlocked = &unlocked;
+            recv.for_each_concurrent(4096, move |incoming| async move {
+                //println!("SRV_RECV: {:?}", incoming);
```
Comment (freesig) [inline crates/lair_keystore_api/src/lair_server/priv_srv.rs:206]: Maybe not needed?

**Review by freesig — APPROVED** [review]
Aproving but a lot of this is over my head.

**Review by maackle — APPROVED** [review]
I like the enum switcheroo for Pending/Running

File: crates/lair_keystore_api/src/lair_server.rs [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+    /// trait object type for AsyncWrite instance.
+    pub type RawSend = Box<dyn tokio::io::AsyncWrite + 'static + Send + Unpin>;
+
+    /// trait object type for AsyncRead instance.
+    pub type RawRecv = Box<dyn tokio::io::AsyncRead + 'static + Send + Unpin>;
+
+    /// host a lair keystore
+    pub trait AsLairServer: 'static + Send + Sync {
+        /// It is much more secure to unlock a lair keystore server
```
Comment (thedavidmeister) [inline crates/lair_keystore_api/src/lair_server.rs:25]: more secure than what?

**Review by thedavidmeister — APPROVED** [review]

File: crates/lair_keystore_api/src/lair_server.rs (hunk 2 above)
Comment (neonphog) [reply crates/lair_keystore_api/src/lair_server.rs:25]: than having a client unlock it over a connection, but this whole locked concept is deleted in https://github.com/holochain/lair/pull/71

**Pull request merged**